utils/wasm-builder/src/wasm_project.rs

```diff
@@ -427,26 +430,51 @@ extern "C" fn metahash() {{
             .as_ref()
             .expect("Run `WasmProject::generate()` first");
 
-        let original_wasm_path = self
-            .target_dir
-            .join(format!("wasm32-unknown-unknown/{}", self.profile))
-            .join(format!("{}.wasm", &file_base_name));
+        let original_wasm_path = self.target_dir.join(format!(
+            "wasm32-unknown-unknown/{}/{file_base_name}.wasm",
+            self.profile
+        ));
 
         fs::create_dir_all(&self.target_dir).context("Failed to create target directory")?;
 
+        let mut wasm_files = vec![];
+
         for pre_processor in &self.pre_processors {
-            let output = pre_processor.pre_process(original_wasm_path.clone())?;
+            for (path, content) in pre_processor.pre_process(original_wasm_path.clone())? {
+                let pre_processed_path = self.wasm_target_dir.join(path);
+                fs::write(&pre_processed_path, content)?;
+
+                if pre_processed_path.extension() == Some(OsStr::new("wasm")) {
+                    let pre_processor_name = pre_processor.name().to_lowercase();
+                    wasm_files.push((
+                        pre_processed_path,
+                        format!("{file_base_name}_{pre_processor_name}"),
+                    ));
+                }
+            }
+        }
 
-            let mut pre_processed_path = self.wasm_target_dir.clone();
-            pre_processed_path.push(output.path);
+        if wasm_files.is_empty() {
```

Comment:
original always should be pushed, why skip it?

or there should be the way for pre-processor to preserve the original compilation artifact under original name

Reply:
done in 454cf15d50072d43d3db7ba8f0395ed5f5a10907